### PR TITLE
[SPARK-24572][SPARKR] "eager execution" for R shell, IDE

### PR DESCRIPTION
**Pull request by adrian555**
## What changes were proposed in this pull request?

Check the `spark.sql.repl.eagerEval.enabled` configuration property in SparkDataFrame `show()` method. If the `SparkSession` has eager execution enabled, the data will be returned to the R client when the data frame is created. So instead of seeing this
```
> df <- createDataFrame(faithful)
> df
SparkDataFrame[eruptions:double, waiting:double]
```
you will see
```
> df <- createDataFrame(faithful)
> df
+---------+-------+                                                             
|eruptions|waiting|
+---------+-------+
|      3.6|   79.0|
|      1.8|   54.0|
|    3.333|   74.0|
|    2.283|   62.0|
|    4.533|   85.0|
|    2.883|   55.0|
|      4.7|   88.0|
|      3.6|   85.0|
|     1.95|   51.0|
|     4.35|   85.0|
|    1.833|   54.0|
|    3.917|   84.0|
|      4.2|   78.0|
|     1.75|   47.0|
|      4.7|   83.0|
|    2.167|   52.0|
|     1.75|   62.0|
|      4.8|   84.0|
|      1.6|   52.0|
|     4.25|   79.0|
+---------+-------+
only showing top 20 rows
```

## How was this patch tested?
Manual tests as well as unit tests (one new test case is added).

### CODE REVIEW

**Review by felixcheung — COMMENTED** [review]
very cool. the core func is good, a few comments on doc, example etc.

File: docs/sparkr.md [hunk 1]
```diff
@@ -450,6 +450,42 @@ print(model.summaries)
 {% endhighlight %}
 </div>
 
+### Eager execution
```
[the remark this answers is not in the document]
Comment (adrian555) [reply docs/sparkr.md]: This will be the same level as "Applying User-Defined Function" where its parent topic is "SparkDataFrame Operations". So the heading is with three `#`.

File: docs/sparkr.md [hunk 2]
```diff
@@ -450,6 +450,42 @@ print(model.summaries)
...
+
+If the eager execution is enabled, the data will be returned to R client immediately when the `SparkDataFrame` is created. Eager execution can be enabled by setting the configuration property `spark.sql.repl.eagerEval.enabled` to `true` when the `SparkSession` is started up.
+
+<div data-lang="r" markdown="1">
+{% highlight r %}
+
+# Start up spark session with eager execution enabled
+sparkR.session(master = "local[*]", sparkConfig = list(spark.sql.repl.eagerEval.enabled = "true"))
+
+df <- createDataFrame(faithful)
```
[the remark this answers is not in the document]
Comment (adrian555) [reply docs/sparkr.md]: Done

File: docs/sparkr.md [hunk 3]
```diff
@@ -450,6 +450,42 @@ print(model.summaries)
...
+
+# Instead of displaying the SparkDataFrame class, displays the data returned
```
[the remark this answers is not in the document]
Comment (adrian555) [reply docs/sparkr.md]: Done

File: docs/sparkr.md [hunk 4]
```diff
@@ -450,6 +450,42 @@ print(model.summaries)
...
+##|      3.6|   85.0|
+##|     1.95|   51.0|
+##|     4.35|   85.0|
+##+---------+-------+
+##only showing top 10 rows
+
+{% endhighlight %} 
+</div>
+
+Note that the `SparkSession` created by `sparkR` shell does not have eager execution enabled. You can stop the current session and start up a new session like above to enable.
```
[the remark this answers is not in the document]
Comment (adrian555) [reply docs/sparkr.md]: Changed to this: 
```
Note that to enable eager execution through `sparkR` command, add `spark.sql.repl.eagerEval.enabled=true` configuration property to the `--conf` option.
```

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+library(testthat)
+
+context("Show SparkDataFrame when eager execution is enabled.")
+
+test_that("eager execution is not enabled", {
```
[the remark this answers is not in the document]
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Since this test will terminate the spark session and restart, I think it is better to be self contained.

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  # Start Spark session without eager execution enabled
+  sparkSession <- if (windows_with_hadoop()) {
+    sparkR.session(master = sparkRTestMaster)
+  } else {
+    sparkR.session(master = sparkRTestMaster, enableHiveSupport = FALSE)
+  }
+  
+  df <- suppressWarnings(createDataFrame(iris))
```
[the remark this answers is not in the document]
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Done

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  # Start Spark session without eager execution enabled
+  sparkSession <- if (windows_with_hadoop()) {
+    sparkR.session(master = sparkRTestMaster,
+                   sparkConfig = list(spark.sql.repl.eagerEval.enabled = "true"))
+  } else {
+    sparkR.session(master = sparkRTestMaster, enableHiveSupport = FALSE, 
+                   sparkConfig = list(spark.sql.repl.eagerEval.enabled = "true"))
+  }
+  
+  df <- suppressWarnings(createDataFrame(iris))
```
[the remark this answers is not in the document]
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Done

File: R/pkg/R/DataFrame.R [hunk 8]
```diff
@@ -244,11 +244,15 @@ setMethod("showDF",
 #' @note show(SparkDataFrame) since 1.4.0
 setMethod("show", "SparkDataFrame",
           function(object) {
-            cols <- lapply(dtypes(object), function(l) {
-              paste(l, collapse = ":")
-            })
-            s <- paste(cols, collapse = ", ")
-            cat(paste(class(object), "[", s, "]\n", sep = ""))
+            if (identical(sparkR.conf("spark.sql.repl.eagerEval.enabled", "false")[[1]], "true")) {
```
[the remark this answers is not in the document]
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: Done

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  expect_is(df, "SparkDataFrame")
+  expected <- "eruptions:double, waiting:double"
+  expect_output(show(df), expected)
+  
+  # Stop Spark session
+  sparkR.session.stop()
+})
+
+test_that("eager execution is enabled", {
+  # Start Spark session without eager execution enabled
```
Comment (dilipbiswal) [inline R/pkg/tests/fulltests/test_eager_execution.R]: Here we are testing with eager execution enabled, right ? Do we need to fix the comment here ?

Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Done

File: R/pkg/R/DataFrame.R [hunk 10]
```diff
@@ -244,11 +245,15 @@ setMethod("showDF",
...
```
Comment (viirya) [inline R/pkg/R/DataFrame.R]: `show` also works on other Spark object like `Column`, I think we only need to do `showDF` for `SparkDataFrame`?

Comment (viirya) [reply R/pkg/R/DataFrame.R]: Or the param doc is not correct?

Comment (adrian555) [reply R/pkg/R/DataFrame.R]: I think the `param` is not correct... the signature of `show()` has `SparkDataFrame` so in this one it won't work for `Column`. In column.R file, it has its own `show()` implementation.

Comment (viirya) [reply R/pkg/R/DataFrame.R]: Can you also correct the param doc?

Comment (adrian555) [reply R/pkg/R/DataFrame.R]: I might have misunderstood your previous question. The `@param` does not need to change as it  is common to `SparkDataFrame` and `Column` etc for `show()` method. Same applies to description, so I rewrote the description. Please have a look.

[the remark this answers is not in the document]
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: Had thought about this. First, I consider it is not in the scope of this jira, because I think they are conflicting with the current `showDF()` behavior. 

Some details: the `showDF()` already takes `numRows` and `truncate` arguments. So if we are going to respect those two as well, we have to decide what behavior is best suitable for `showDF()`. For example, whether `showDF()` should just ignore the eager execution, or it picks the `maxNumRows` and `truncate` set through eager execution like following:

```
setMethod("showDF",
          signature(x = "SparkDataFrame"),
          function(x, numRows = 20, truncate = TRUE, vertical = FALSE) {
            eagerNumRows <- as.numeric(sparkR.conf("spark.sql.repl.eagerEval.maxNumRows", "0")[[1]])
            numRows <- ifelse(eagerNumRows == 0, numRows, eagerNumRows)
            eagerTruncate <- as.numeric(sparkR.conf("spark.sql.repl.eagerEval.truncate", "0")[[1]])
            truncate <- ifelse(eagerTruncate == 0, truncate, eagerTruncate)
            if (is.logical(truncate) && truncate) {
              s <- callJMethod(x@sdf, "showString", numToInt(numRows), numToInt(20), vertical)
            } else {
              truncate2 <- as.numeric(truncate)
              s <- callJMethod(x@sdf, "showString", numToInt(numRows), numToInt(truncate2),
                               vertical)
            }
            cat(s)
          })
```

And if we think that `showDF()` can ignore the eager execution setting and still want the `show()` to observe eager execution config, we can certainly just grab the `maxNumRows` and `truncate` setting and pass to `showDF() call.

However, my second point is that I don't think these two configs matter much or that important/necessary. Since the eager execution is just to show a snippet data of the SparkDataFrame, our default `numRows = 20` and `truncate = TRUE` are good enough iMO. If users want to see more or less number of rows, they should call `showDF()`.

@felixcheung, your thought?

[the remark this answers is not in the document]
Comment (dilipbiswal) [reply R/pkg/R/DataFrame.R]: @adrian555 Thanks for the explanation. 
> However, my second point is that I don't think these two configs matter much or that important/necessary. Since the eager execution is just to show a snippet data of the SparkDataFrame, our default numRows = 20 and truncate = TRUE are good enough iMO. If users want to see more or less number of rows, they should call showDF().

So i just wanted to make sure if its possible to have parity with how it works for python. It seems to me that in python, we just get the two configs and call the showstring method.

> And if we think that showDF() can ignore the eager execution setting and still want the show() to observe eager execution config, we can certainly just grab the maxNumRows and truncate setting and pass to showDF() call.

What will happen if we grab these config in show() when eager execution is enabled and then call showDF() by passing these parameters ?

Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: ^ that - it's two different "overloads" for two different "class"

[the remark this answers is not in the document]
Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: respecting `spark.sql.repl.eagerEval.maxNumRows` somewhat makes sense. but instead of changing `showDF` which has other cases beyond eagerEval, we could change where `showDF` is called by `show`, and pass the max rows.

here https://github.com/apache/spark/pull/22455/files#diff-508641a8bd6c6b59f3e77c80cdcfa6a9R249

**Comment by viirya** [comment]
Let's also update the doc of `REPL_EAGER_EVAL_ENABLED` in `SQLConf`. After this patch, eager evaluation is not only supported in PySpark.

File: R/pkg/R/DataFrame.R [hunk 11]
```diff
@@ -226,7 +226,8 @@ setMethod("showDF",
 
 #' show
 #'
-#' Print class and type information of a Spark object.
+#' If eager evaluation is enabled and the Spark object is a SparkDataFrame, return the data of
```
Comment (viirya) [inline R/pkg/R/DataFrame.R]: return the data of the SparkDataFrame object -> evaluate the SparkDataFrame and print top rows of the SparkDataFrame

File: R/pkg/R/DataFrame.R (hunk 10 above)
[the remark this answers is not in the document]
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: Done

File: R/pkg/R/DataFrame.R (hunk 11 above)
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: Done

**Comment by dilipbiswal** [comment]
@adrian555 The changes look fine to me. Thank you.

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+test_that("eager execution is enabled", {
+  # Start Spark session with eager execution enabled
+  sparkSession <- if (windows_with_hadoop()) {
+    sparkR.session(master = sparkRTestMaster,
+                   sparkConfig = list(spark.sql.repl.eagerEval.enabled = "true",
+                                      spark.sql.repl.eagerEval.maxNumRows = as.integer(10)))
+  } else {
+    sparkR.session(master = sparkRTestMaster, enableHiveSupport = FALSE, 
+                   sparkConfig = list(spark.sql.repl.eagerEval.enabled = "true",
+                                      spark.sql.repl.eagerEval.maxNumRows = as.integer(10)))
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_eager_execution.R]: I'd set sparkConfig to a list before this and simply reuse it for both cases instead of having them duplicated

File: R/pkg/R/DataFrame.R [hunk 13]
```diff
@@ -244,11 +246,25 @@ setMethod("showDF",
 #' @note show(SparkDataFrame) since 1.4.0
 setMethod("show", "SparkDataFrame",
           function(object) {
-            cols <- lapply(dtypes(object), function(l) {
-              paste(l, collapse = ":")
-            })
-            s <- paste(cols, collapse = ", ")
-            cat(paste(class(object), "[", s, "]\n", sep = ""))
+            if (identical(sparkR.conf("spark.sql.repl.eagerEval.enabled", "false")[[1]], "true")) {
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: instead of calling sparkR.conf three times, I'm wondering if it cleaner/safer to get a copy like in here:
https://github.com/apache/spark/blob/95b177c8f0862c6965a7c3cd76b3935c975adee9/R/pkg/tests/fulltests/test_sparkSQL.R#L3538

File: R/pkg/R/DataFrame.R [hunk 14]
```diff
@@ -244,11 +246,25 @@ setMethod("showDF",
...
+              argsList <- list()
+              argsList$x <- object
+              numRows <- as.numeric(sparkR.conf("spark.sql.repl.eagerEval.maxNumRows", "0")[[1]])
+              if (numRows > 0) {
+                argsList$numRows <- numRows
+              }
+              truncate <- as.numeric(sparkR.conf("spark.sql.repl.eagerEval.truncate", "0")[[1]])
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: have you tested with the case when only `spark.sql.repl.eagerEval.truncate` is set, and `spark.sql.repl.eagerEval.maxNumRows` is not?

File: docs/sparkr.md [hunk 15]
```diff
@@ -450,6 +450,48 @@ print(model.summaries)
 {% endhighlight %}
 </div>
 
+### Eager execution
+
+If the eager execution is enabled, the data will be returned to R client immediately when the `SparkDataFrame` is created. Eager execution can be enabled by setting the configuration property `spark.sql.repl.eagerEval.enabled` to `true` when the `SparkSession` is started up.
+
+Maximum number of rows and maximum number of characters per column of data to display can be controlled by `spark.sql.repl.eagerEval.maxNumRows` and `spark.sql.repl.eagerEval.truncate` configuration properties, respectively.
```
Comment (HyukjinKwon) [inline docs/sparkr.md]: Let's note the default values.

File: R/pkg/tests/fulltests/test_eager_execution.R (hunk 12 above)
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Done

File: R/pkg/R/DataFrame.R (hunk 13 above)
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: Retrieving all conf through `sparkR.conf()` does not provide a default value for each conf, so extra condition check is needed. Ideally, I believe we should have cached all configuration for the session inside the sparkr env to avoid backend calls at all.

Anyway, I made the suggested change.

File: docs/sparkr.md (hunk 15 above)
Comment (adrian555) [reply docs/sparkr.md]: Done

File: R/pkg/R/DataFrame.R (hunk 14 above)
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: I did test and it worked. Any specific concern? Thanks.

**Comment by adrian555** [comment]
Looks like the test was terminated in the middle, not likely related to the any code change. Could someone please ask for retest? Thanks.

**Comment by gatorsmile** [comment]
Great feature! cc @falaki

**Comment by felixcheung** [comment]
> Looks like the test was terminated in the middle, not likely related to the any code change. Could someone please ask for retest? Thanks.

appveyor one? you need to close and re-open this PR to trigger it

File: R/pkg/R/DataFrame.R [hunk 16]
```diff
@@ -244,11 +246,31 @@ setMethod("showDF",
 #' @note show(SparkDataFrame) since 1.4.0
 setMethod("show", "SparkDataFrame",
           function(object) {
-            cols <- lapply(dtypes(object), function(l) {
-              paste(l, collapse = ":")
-            })
-            s <- paste(cols, collapse = ", ")
-            cat(paste(class(object), "[", s, "]\n", sep = ""))
+            allConf <- sparkR.conf()
+            if (!is.null(allConf[["spark.sql.repl.eagerEval.enabled"]]) &&
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: you can also save off `foo <- allConf[["spark.sql.repl.eagerEval.enabled"]]` and reuse it, ditto for other cases

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 17]
```diff
@@ -1491,9 +1491,10 @@ object SQLConf {
   val REPL_EAGER_EVAL_ENABLED = buildConf("spark.sql.repl.eagerEval.enabled")
     .doc("Enables eager evaluation or not. When true, the top K rows of Dataset will be " +
       "displayed if and only if the REPL supports the eager evaluation. Currently, the " +
-      "eager evaluation is only supported in PySpark. For the notebooks like Jupyter, " +
-      "the HTML table (generated by _repr_html_) will be returned. For plain Python REPL, " +
-      "the returned outputs are formatted like dataframe.show().")
+      "eager evaluation is supported in PySpark and SparkR. In PySpark, for the notebooks like " +
+      "Jupyter, the HTML table (generated by _repr_html_) will be returned. For plain Python " +
+      "REPL, the returned outputs are formatted like dataframe.show(). In SparkR, the returned " +
+      "outputs are showed as R data.frame.")
```
Comment (felixcheung) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ` are showed as R data.frame` - it's not actually the exact format, so how about we say ` are showed similar to R data.frame would`

File: docs/sparkr.md [hunk 18]
```diff
@@ -450,6 +450,48 @@ print(model.summaries)
...
+##|   51.0|    6|
+##|   52.0|    5|
+##|   53.0|    7|
+##+-------+-----+
+##only showing top 10 rows
+
+{% endhighlight %} 
+</div>
+
+Note that to enable eager execution through `sparkR` command, add `spark.sql.repl.eagerEval.enabled=true` configuration property to the `--conf` option.
```
Comment (felixcheung) [inline docs/sparkr.md]: `sparkR command` - I think should be `sparkR shell`?

File: R/pkg/R/DataFrame.R (hunk 14 above)
Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: just asking since this variation was not in unit test

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+library(testthat)
+
+context("Show SparkDataFrame when eager execution is enabled.")
+
+test_that("eager execution is not enabled", {
+  # Start Spark session without eager execution enabled
+  sparkSession <- if (windows_with_hadoop()) {
+    sparkR.session(master = sparkRTestMaster)
+  } else {
+    sparkR.session(master = sparkRTestMaster, enableHiveSupport = FALSE)
```
[the remark this answers is not in the document]
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Done

File: R/pkg/R/DataFrame.R (hunk 16 above)
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 17 above)
Comment (adrian555) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Done

File: docs/sparkr.md (hunk 18 above)
Comment (adrian555) [reply docs/sparkr.md]: In the same doc "From Data Sources", it has `either be added by specifying --packages with spark-submit or sparkR commands`, that is why I used `command` instead of `shell`. I would think that `script`, `shell` and `command` are exchangeable here. But if viewed by the angle that `sparkR` ends with a R execution env, maybe `shell` makes more sense. :)

So I made the change.

**Comment by falaki** [comment]
@adrian555 thanks for submitting this. Can we have a config to set the default `print` function in eager mode. It can default to `show`, but I can imagine it is useful to make it configurable. For example in Jupyter there are ricer ways of displaying R data.frame.

**Comment by adrian555** [comment]
@falaki trying to understand what you were asking for. I know that `print` takes some arguments like `digit` and `quote` that can further format the output. `print` is for S3 object and `show` is for S4 object. So, are you asking for `print.SparkDataFrame` function as follow
```
print.SparkDataFrame <- function(x, ...) {
  argsList <- list(...)
  if (!is.null(argsList$eager) && !argsList$eager) {
    cols <- lapply(dtypes(x), function(l) {
      paste(l, collapse = ":")
    })
    s <- paste(cols, collapse = ", ")
    cat(paste(class(x), "[", s, "]\n", sep = ""))
  } else {
    show(x)
  }
}
```
With this, 
* `print(df)` will have the same behavior as `show(df)`. 
* `print(df, eager=FALSE)` will continue to display just the SparkDataFrame class info, even eager execution is enabled. 
* `print(df, eager=TRUE)` will be the same as `show()` when eager execution is enabled. But if eager execution is disable, the passed in `eager=TRUE` is ignored. (This way, we guarantee only the session's SQLConf is in effect.)

Also, are you also thinking of passing the `maxNumRows` and `truncate` as arguments to `print.SparkDataFrame` and replacing `spark.sql.repl.eagerEval.maxNumRows` and `spark.sql.repl.eagerEval.truncate`?

Please let me know if this is on par with your thought then I will push a commit.

Thanks.

**Comment by adrian555** [comment]
Another flavor will be this
```r
print.SparkDataFrame <- function(x, ...) {
  if ((length(list(...)) > 0) && identical(sparkR.conf("spark.sql.repl.eagerEval.enabled", "false")[[1]], "true")) {
    # TODO: limit the maxNumRows and truncate
    df <- collect(x)
    print(df, ...)
  } else {
    show(x)
  }
}
```
this allows the other `print` formats to be taken. For example, when eager execution is enabled, `print(createDataFrame(iris), quote=TRUE)` will have such output
```
    Sepal_Length Sepal_Width Petal_Length Petal_Width      Species              
1          "5.1"       "3.5"        "1.4"       "0.2"     "setosa"
2          "4.9"       "3.0"        "1.4"       "0.2"     "setosa"
3          "4.7"       "3.2"        "1.3"       "0.2"     "setosa"
4          "4.6"       "3.1"        "1.5"       "0.2"     "setosa"
```

**Comment by falaki** [comment]
@adrian555 These all great points. My high-level was enabling other platforms (e.g., Jupyter) to plugin more advanced (custom) functions for displaying SparkDataFrame. If the framework does not set anything, SparkR would use `show()` by default. But if the notebook environment has a nicer way of displaying, they can simply set a config to use that by default.

**Comment by adrian555** [comment]
Thanks @falaki. We are taking out the `print` method out so the requirement seems clear now. Yes, for Jupyter with IRkernel, two options `jupyter.rich_display` and jupyter.base_display_func` are used to display the data.frame in a nicer way, as you mentioned. And since it is a kernel with a backend, the automatic print is not really calling the `show` method. Instead, the data is evaluated and the resulted message is returned as a response. Hence, the `df` and `show(df)` have different output formats.

I think you want to have a common option in SparkR that other framework can use it instead of creating its own. Will `sparkr.SparkDataFrame.base_show_func` work for you? And I will modify the `show()` method for `SparkDataFrame` as follow
```r
setMethod("show", "SparkDataFrame",
          function(object) {
            showFunc <- getOption("sparkr.SparkDataFrame.base_show_func")
            if (!is.null(showFunc)) {
              stopifnot(typeof(showFunc) == "closure")
              showFunc(object)
            } else {
              allConf <- sparkR.conf()
              prop <- allConf[["spark.sql.repl.eagerEval.enabled"]]
              if (!is.null(prop) && identical(prop, "true")) {
                argsList <- list()
                argsList$x <- object
                prop <- allConf[["spark.sql.repl.eagerEval.maxNumRows"]]
                if (!is.null(prop)) {
                  numRows <- as.numeric(prop)
                  if (numRows > 0) {
                    argsList$numRows <- numRows
                  }
                }
                prop <- allConf[["spark.sql.repl.eagerEval.truncate"]]
                if (!is.null(prop)) {
                  truncate <- as.numeric(prop)
                  if (truncate > 0) {
                    argsList$truncate <- truncate
                  }
                }
                do.call(showDF, argsList)
              } else {
                cols <- lapply(dtypes(object), function(l) {
                  paste(l, collapse = ":")
                })
                s <- paste(cols, collapse = ", ")
                cat(paste(class(object), "[", s, "]\n", sep = ""))
              }              
            }
          })
```
Basically if that option is set to a predefined function, it will run that function to show. This approach also makes the `df` and `show(df)` act the same.

Please have a look. Thanks.

**Comment by falaki** [comment]
@adrian555 yes, that looks good. Thank you!

File: R/pkg/R/DataFrame.R [hunk 20]
```diff
@@ -246,30 +248,38 @@ setMethod("showDF",
...
-            allConf <- sparkR.conf()
-            if (!is.null(allConf[["spark.sql.repl.eagerEval.enabled"]]) &&
-                identical(allConf[["spark.sql.repl.eagerEval.enabled"]], "true")) {
-              argsList <- list()
-              argsList$x <- object
-              if (!is.null(allConf[["spark.sql.repl.eagerEval.maxNumRows"]])) {
-                numRows <- as.numeric(allConf[["spark.sql.repl.eagerEval.maxNumRows"]])
-                if (numRows > 0) {
-                  argsList$numRows <- numRows
+            showFunc <- getOption("sparkr.SparkDataFrame.base_show_func")
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: hmm, this naming convention? typically we don't mix `.` and `_` and I don't think we have anything with `SparkDataFrame`

seems like we have `spark.sparkr.something` before https://github.com/apache/spark/blob/5264164a67df498b73facae207eda12ee133be7d/core/src/main/scala/org/apache/spark/api/r/RRunner.scala#L343

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 21]
```diff
@@ -21,11 +21,7 @@ context("Show SparkDataFrame when eager execution is enabled.")
 
 test_that("eager execution is not enabled", {
   # Start Spark session without eager execution enabled
-  sparkSession <- if (windows_with_hadoop()) {
-    sparkR.session(master = sparkRTestMaster)
-  } else {
-    sparkR.session(master = sparkRTestMaster, enableHiveSupport = FALSE)
-  }
+  sparkR.session(master = sparkRTestMaster)
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_eager_execution.R]: you should definitely set `enableHiveSupport = FALSE` - historically this hasn't work well in other R tests when hive catalog is enabled

Comment (felixcheung) [reply R/pkg/tests/fulltests/test_eager_execution.R]: as mentioned here https://github.com/apache/spark/pull/22455#discussion_r220030686

**Review by felixcheung — COMMENTED** [review]
I'm not sure about the new changes. IMO we are mixing S3 and S4 convention here as per @adrian555.

IMO, we should have an option for show to return a data.frame, if we are running in an IDE. then we can leverage any custom print logic the IDE has
````
print.data.frame
function (x, ..., digits = NULL, quote = FALSE, right = TRUE,
    row.names = TRUE)
```
?

File: R/pkg/tests/fulltests/test_eager_execution.R (hunk 21 above)
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Sure. Guess I misinterpreted your other comment.

File: R/pkg/R/DataFrame.R (hunk 20 above)
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: I don't have any naming preference. One thing to note is that the `spark.sparkr.r.command` you mentioned above is the config to the Spark while the one I tried to define here is a config in the R side.

**Comment by adrian555** [comment]
> I'm not sure about the new changes. IMO we are mixing S3 and S4 convention here as per @adrian555.

The implemented code is still conforming to S4 object standard that `show()` is the method to display the object.

> IMO, we should have an option for show to return a data.frame, if we are running in an IDE. then we can leverage any custom print logic the IDE has

`show()` ideally should return an invisible NULL to be consistent with other R objects.

CC @falaki for any further discussion on this topic.

File: R/pkg/R/DataFrame.R (hunk 20 above)
Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: IMO pretty print should plug in to something more R standard like 
[printr](https://yihui.name/printr/) or
[lemon](https://cran.r-project.org/web/packages/lemon/vignettes/lemon_print.html)
or
[print.x](https://stat.ethz.ch/R-manual/R-devel/library/base/html/print.dataframe.html)

Comment (felixcheung) [reply R/pkg/R/DataFrame.R]: could we consider leaving print/show option out? I'd like to get eager compute to work even in basic sparkR / R shell

Comment (adrian555) [reply R/pkg/R/DataFrame.R]: It seems a good idea to separate out the print/show issue as I can see it is not 100% related to this jira. Maybe it will help with a new jira to track the requirement and discussion for that issue. Please confirm so that I will revert some related changes. Thanks.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: R/pkg/R/DataFrame.R (hunk 20 above)
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: I have backed out the change of display function option. Also opened jira https://issues.apache.org/jira/browse/SPARK-25770 to track that issue.

File: R/pkg/R/DataFrame.R [hunk 22]
```diff
@@ -244,11 +246,33 @@ setMethod("showDF",
...
+            allConf <- sparkR.conf()
+            prop <- allConf[["spark.sql.repl.eagerEval.enabled"]]
+            if (!is.null(prop) && identical(prop, "true")) {
+              argsList <- list()
+              argsList$x <- object
+              prop <- allConf[["spark.sql.repl.eagerEval.maxNumRows"]]
+              if (!is.null(prop)) {
+                numRows <- as.numeric(prop)
+                if (numRows > 0) {
+                  argsList$numRows <- numRows
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: could you manually check if this works correctly when numRows or truncate (L265) is a numeric but not an integer - eg. 1.4

if not, we should replace `as.numeric` than `as.integer` in both places

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+  expect_output(show(df), expected)
+  
+  # Stop Spark session
+  sparkR.session.stop()
+})
+
+test_that("eager execution is enabled", {
+  # Start Spark session with eager execution enabled
+  sparkConfig <- list(spark.sql.repl.eagerEval.enabled = "true",
+                      spark.sql.repl.eagerEval.maxNumRows = as.integer(10))
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_eager_execution.R]: add a test with `spark.sql.repl.eagerEval.maxNumRows` not set.
change this test to add `spark.sql.repl.eagerEval.truncate`

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 24]
```diff
@@ -3731,6 +3731,7 @@ test_that("catalog APIs, listTables, listColumns, listFunctions", {
   dropTempView("cars")
 })
 
+
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: remove empty line

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 25]
```diff
@@ -0,0 +1,52 @@
+#
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_eager_execution.R]: one thing to note, since test runs alphebatically, this test will run before sparkSQL - I think we should rename this to test_sparkSQL_eager.R to ensure it runs after

File: R/pkg/tests/fulltests/test_eager_execution.R [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+library(testthat)
+
+context("Show SparkDataFrame when eager execution is enabled.")
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_eager_execution.R]: change to `test show SparkDataFrame when eager execution is enabled.`

**Review by felixcheung — CHANGES_REQUESTED** [review]
a few comment, please update and we would be good to merge

File: R/pkg/R/DataFrame.R (hunk 22 above)
Comment (adrian555) [reply R/pkg/R/DataFrame.R]: A spark session would not start if the maxNumRows config is not an integer. So replacing as.numeric with as.integer here won't make real difference. But I made the code change anyway to be clearer from code view.

File: R/pkg/tests/fulltests/test_eager_execution.R (hunk 25 above)
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Done.

File: R/pkg/tests/fulltests/test_eager_execution.R (hunk 26 above)
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Done

File: R/pkg/tests/fulltests/test_eager_execution.R (hunk 23 above)
Comment (adrian555) [reply R/pkg/tests/fulltests/test_eager_execution.R]: Done

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 24 above)
Comment (adrian555) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Done

**Comment by adrian555** [comment]
Test was terminated as "Build execution time has reached the maximum allowed time for your plan (90 minutes)."

Try to close and reopen the PR to see how the retest goes.

**Comment by felixcheung** [comment]
LGTM. merged to master
